mypy/types.py

```diff
@@ -1394,3 +1419,62 @@ def is_named_instance(t: Type, fullname: str) -> bool:
     return (isinstance(t, Instance) and
             t.type is not None and
             t.type.fullname() == fullname)
+
+
+def copy_type(t: Type) -> Type:
+    """
+    Build a copy of the type; used to mutate the copy with truthiness information
+    """
+    return copy.copy(t)
+
+
+def true_only(t: Type) -> Type:
+    """
+    Restricted version of t with only True-ish values
+    """
+    if not t.can_be_true:
+        # All values of t are False-ish, so there are no true values in it
+        return UninhabitedType(line=t.line)
+    elif not t.can_be_false:
+        # All values of t are already True-ish, so true_only is idempotent in this case
+        return t
+    elif isinstance(t, UnionType):
+        # The true version of a union type is the union of the true versions of its components
+        new_items = [true_only(item) for item in t.items]
+        return UnionType.make_simplified_union(new_items, line=t.line)
+    else:
+        new_t = copy_type(t)
+        new_t.can_be_false = False
+        return new_t
+
+
+def false_only(t: Type) -> Type:
+    """
+    Restricted version of t with only False-ish values
+    """
+    if not t.can_be_false:
+        # All values of t are True-ish, so there are no false values in it
+        return UninhabitedType(line=t.line)
+    elif not t.can_be_true:
+        # All values of t are already False-ish, so false_only is idempotent in this case
+        return t
+    elif isinstance(t, UnionType):
+        # The false version of a union type is the union of the false versions of its components
+        new_items = [false_only(item) for item in t.items]
+        return UnionType.make_simplified_union(new_items, line=t.line)
+    else:
+        new_t = copy_type(t)
+        new_t.can_be_true = False
+        return new_t
+
+
+def true_or_false(t: Type) -> Type:
+    """
+    Unrestricted version of t with both True-ish and False-ish values
+    """
+    new_t = copy_type(t)
+    if not new_t.can_be_true and type(new_t).can_be_true:
```

Comment:
I think this might be a bit simpler with an unconditional `new_t.can_be_true = type(new_t).can_be_true`.